src/fractionatedmorse.rs

```diff
@@ -0,0 +1,333 @@
+                }
+            }
+            // Add a single space between each word. Spaces are represented having nothing
+            // between two seperators, `||`, so we push another seperator here. This also takes
+            // care of the final `||` required to end the message.
+            morse.push('|');
+        }
+
+        Ok(morse)
+    }
+
+    /// Takes a Morse code string, with each Morse character separated by `|`, and converts it to
+    /// plaintext. This function returns `Err` if an invalid Morse character is encountered.
+    fn decrypt_morse(mut message: String) -> Result<String, &'static str> {
+        let mut plaintext = String::new();
+
+        // Remove character separators from the beginning of the message if present
+        while message.starts_with('|') {
+            message.remove(0);
+        }
+
+        // Loop over every Morse character
+        for morse_chr in message.split('|') {
+            // Find the Morse character in the alphabet and decode it.
+            if let Some(pos) = MORSE_ALPHABET.iter().position(|&m| m.1 == morse_chr) {
+                plaintext.push(MORSE_ALPHABET[pos].0);
+            } else {
+                return Err("Invalid Fractionated Morse message. Unknown Morse character found.")
+            }
+        }
+
+        // Messages are ended with `||` but this will be decoded as a two spaces in the loop above.
+        // Therefore remove any trailing whitespace. This will also remove any whitespace present
+        // at the end of the plaintext.
+        plaintext = plaintext.trim_right().to_string();
+
+        Ok(plaintext)
+    }
+
+    /// Takes a alphabetic string and converts it to Fractionated Morse. This function will
+    /// return `Err` if a non-alphabetic symbol is present in the message.
+    fn encrypt_frac_morse(keyed_alphabet: &String, message: String) -> Result<String, &'static str> {
+        let mut frac_morse = String::new();
+
+        // We are using a keyed alphabet which is lowercase, therefore loop over a lowercase version
+        // of the message.
+        for c in message.to_lowercase().chars() {
+            if let Some(pos) = keyed_alphabet.chars().position(|a| a == c) {
+                frac_morse.extend(FRAC_MORSE_ALPHABET[pos].chars());
+            } else {
+                return Err("Invalid message. Please strip any whitespace or non-alphabetic symbols.")
+            }
+        }
+
+        Ok(frac_morse)
+    }
+
+    /// Takes a Morse string, pads it with dots to a length that is a multiple of 3, and converts
+    /// it to an alphabetic string. This function returns `Err` if an invalid Fractionated Morse
+    /// trigraph is encountered.
+    fn decrypt_frac_morse(keyed_alphabet: &String, mut message: String) -> Result<String, &'static str> {
+        let mut ciphertext = String::new();
+
+        // Trigraphs are required, so pad the message with dots until the length is a multiple of 3.
+        while message.len() % 3 != 0 {

```

Comment:
I think rather than making `message` a mutable variable in this function, this padding code should be moved to its own function.